glean-core/src/event_database/mod.rs

```diff
@@ -0,0 +1,375 @@
+    ///   `events` will be created inside of this directory.
+    pub fn new(data_path: &str) -> Result<Self> {
+        let path = Path::new(data_path).join("events");
+        create_dir_all(&path)?;
+
+        Ok(Self {
+            path,
+            event_stores: RwLock::new(HashMap::new()),
+        })
+    }
+
+    /// Initialize events storage. This must be called once on application startup,
+    /// e.g. from [Glean.initialize], but after we are ready to send pings, since
+    /// this could potentially collect and send pings.
+    ///
+    /// If there are any events queued on disk, it loads them into memory so
+    /// that the memory and disk representations are in sync.
+    ///
+    /// Secondly, if this is the first time the application has been run since
+    /// rebooting, any pings containing events are assembled into pings and cleared
+    /// immediately, since their timestamps won't be compatible with the timestamps
+    /// we would create during this boot of the device.
+    ///
+    /// # Arguments
+    ///
+    /// * `glean`:  The Glean instance.
+    pub fn on_ready_to_send_pings(&self, glean: &Glean) {
+        match self.load_events_from_disk() {
+            Ok(_) => self.send_all_events(glean),
+            Err(err) => log::error!("Error loading pings from disk: {}", err),
+        }
+    }
+
+    fn load_events_from_disk(&self) -> Result<()> {
+        let mut db = self.event_stores.write().unwrap();
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            if entry.file_type()?.is_file() {
+                let store_name = entry.file_name().into_string()?;
+                let file = BufReader::new(OpenOptions::new().read(true).open(entry.path())?);
+                db.insert(
+                    store_name,
+                    file.lines()
+                        .filter_map(|line| line.ok())
+                        .filter_map(|line| {
+                            println!("{}", line);
+                            serde_json::from_str::<JsonValue>(&line).ok()
+                        })
+                        .filter_map(|json| RecordedEventData::deserialize(&json))
+                        .collect(),
+                );
+            }
+        }
+        Ok(())
+    }
+
+    fn send_all_events(&self, glean: &Glean) {
+        let store_names = {
+            let db = self.event_stores.read().unwrap();
+            db.keys().cloned().collect::<Vec<String>>()
+        };
+
+        for store_name in store_names {
+            if let Err(err) = glean.send_ping_by_name(&store_name, false) {
+                log::error!(
+                    "Error flushing existing events to the {} ping: {}",
+                    store_name,
+                    err
+                );
+            }
+        }
+    }
+
+    /// Record an event in the desired stores.
+    ///
+    /// # Arguments
+    ///
+    /// - `glean`: The Glean instance.
```

Comment:
```suggestion
    /// * `glean` - The Glean instance.
```

That's the formatting we use in other places. applies to the following lines as well.